### PR TITLE
Initial Costmap_2d port

### PR DESCRIPTION
**Pull request by bpwilcox**
This initial port of costmap_2d compiles successfully. There still remain many TODO which are documented throughout the code. I'll file issues for these TODO and remaining work.

### CODE REVIEW

File: src/libs/costmap_2d/CMakeLists.txt [hunk 1]
```diff
@@ -1,202 +1,161 @@
...
+remove_definitions(-DDISABLE_LIBUSB-1.0)
 find_package(Eigen3 REQUIRED)
 find_package(Boost REQUIRED COMPONENTS system thread)
+
 include_directories(
-    include
-    ${catkin_INCLUDE_DIRS}
-    ${EIGEN3_INCLUDE_DIRS}
-    ${Boost_INCLUDE_DIRS}
+	include
```
Comment (mjeronimo) [inline src/libs/costmap_2d/CMakeLists.txt]: We've been using 2 space indentation.

File: src/libs/costmap_2d/CMakeLists.txt [hunk 2]
```diff
@@ -1,202 +1,161 @@
...
+  tf2_sensor_msgs
+  xmlrpcpp
+  pluginlib
+  tf2_geometry_msgs
 )
-add_dependencies(costmap_2d ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORTED_TARGETS})
-target_link_libraries(costmap_2d
+
+ament_target_dependencies(nav2_costmap_2d 
+${dependencies}
```
Comment (mjeronimo) [inline src/libs/costmap_2d/CMakeLists.txt]: Two space indentation (should be consistent with the rest of this and similar files).

File: src/libs/costmap_2d/src/costmap_2d_node.cpp [hunk 3]
```diff
@@ -35,18 +35,21 @@
...
 #include <tf2_ros/transform_listener.h>
 
-int main(int argc, char** argv)
+int main(int argc, char ** argv)
 {
-  ros::init(argc, argv, "costmap_node");
-  tf2_ros::Buffer buffer(ros::Duration(10));
+  rclcpp::init(argc, argv);
+
+  auto node = rclcpp::Node::make_shared("costmap_2d");
```
Comment (crdelsey) [inline src/libs/costmap_2d/src/costmap_2d_node.cpp]: This is called *costmap_node* in the original

File: src/libs/costmap_2d/src/costmap_2d_node.cpp [hunk 4]
```diff
@@ -35,18 +35,21 @@
...
+  std::string name = (argc == 1) ? "costmap" : std::string(argv[1]);
```
Comment (crdelsey) [inline src/libs/costmap_2d/src/costmap_2d_node.cpp]: I don't think we should do this. There will likely be other arguments on the command line. Just leave the code to work as it used to.

File: src/libs/costmap_2d/src/costmap_2d_ros.cpp [hunk 5]
```diff
@@ -41,65 +41,76 @@
...
 
 namespace costmap_2d
 {
 
-void move_parameter(ros::NodeHandle& old_h, ros::NodeHandle& new_h, std::string name, bool should_delete = true)
+void move_parameter(rclcpp::Node::SharedPtr old_h, rclcpp::Node::SharedPtr new_h, std::string name,
+    bool should_delete = true)
 {
-  if (!old_h.hasParam(name))
+  auto parameters_client_old = std::make_shared<rclcpp::SyncParametersClient>(old_h);
```
Comment (crdelsey) [inline src/libs/costmap_2d/src/costmap_2d_ros.cpp]: You don't need to make a parameters client. There is one built into the node that is exposed through the node interface.

This block could be something like
```c++
XmlRpc::XmlRpcValue value;
if (!old_h.get_parameter(name, value))
  return;
else
 new_h.set_parameters({rclcpp::Parameter(name, value)});
```

The delete part I still don't know how to do.

File: src/libs/costmap_2d/src/costmap_2d_ros.cpp [hunk 6]
```diff
@@ -41,65 +41,76 @@
...
-
-  ros::Time last_error = ros::Time::now();
+  auto private_parameters_client = std::make_shared<rclcpp::SyncParametersClient>(private_nh);
+  global_frame_ =
+      private_parameters_client->get_parameter<std::string>("global_frame", std::string("map"));
+  robot_base_frame_ =
+      private_parameters_client->get_parameter<std::string>("robot_base_frame", std::string(
+        "base_link"));
+
+  auto clock = new rclcpp::Clock();
```
Comment (crdelsey) [inline src/libs/costmap_2d/src/costmap_2d_ros.cpp]: This is a 'sort of' memory leak. You never delete this pointer. Just make it a local variable
```c++
rclcpp::Clock clock;
rclcpp::Time last_error = clock.now();
```

File: src/libs/costmap_2d/src/costmap_2d_ros.cpp [hunk 7]
```diff
@@ -108,133 +119,146 @@ Costmap2DROS::Costmap2DROS(const std::string& name, tf2_ros::Buffer& tf) :
 
   // check if we want a rolling window version of the costmap
   bool rolling_window, track_unknown_space, always_send_full_costmap;
-  private_nh.param("rolling_window", rolling_window, false);
-  private_nh.param("track_unknown_space", track_unknown_space, false);
-  private_nh.param("always_send_full_costmap", always_send_full_costmap, false);
+
+  rolling_window = private_parameters_client->get_parameter<bool>("rolling_window", false);
```
Comment (crdelsey) [inline src/libs/costmap_2d/src/costmap_2d_ros.cpp]: When you get rid of the parameters client, these would become
```c++
nh->get_parameter_or("rolling_window", rolling_window, false);
...
```

File: src/libs/costmap_2d/src/costmap_2d_ros.cpp [hunk 8]
```diff
@@ -401,47 +436,49 @@ void Costmap2DROS::mapUpdateLoop(double frequency)
...
+          r.period());
+    }
   }
 }
 
 void Costmap2DROS::updateMap()
 {
-  if (!stop_updates_)
-  {
+  auto clock = new rclcpp::Clock();
```
Comment (crdelsey) [inline src/libs/costmap_2d/src/costmap_2d_ros.cpp]: Change to local variable as described above.

File: src/libs/costmap_2d/src/costmap_2d_ros.cpp [hunk 9]
```diff
@@ -495,74 +533,81 @@ void Costmap2DROS::resume()
...
       ++plugin)
   {
     (*plugin)->reset();
   }
 }
 
-bool Costmap2DROS::getRobotPose(geometry_msgs::PoseStamped& global_pose) const
+bool Costmap2DROS::getRobotPose(geometry_msgs::msg::PoseStamped & global_pose) const
 {
+  auto clock = new rclcpp::Clock();
```
Comment (crdelsey) [inline src/libs/costmap_2d/src/costmap_2d_ros.cpp]: Change to local variable as described above.

File: src/libs/costmap_2d/include/costmap_2d/costmap_2d_ros.h [hunk 10]
```diff
@@ -225,12 +232,12 @@ class Costmap2DROS
    * layered_costmap_->setFootprint().  Also saves the unpadded
    * footprint, which is available from
    * getUnpaddedRobotFootprint(). */
-  void setUnpaddedRobotFootprintPolygon(const geometry_msgs::Polygon& footprint);
+  void setUnpaddedRobotFootprintPolygon(const geometry_msgs::msg::Polygon::SharedPtr footprint);
```
[the remark this answers is not in the document]
Comment (bpwilcox) [reply src/libs/costmap_2d/include/costmap_2d/costmap_2d_ros.h]: I was following the examples I've seen for ros2 subscription callbacks which have it as a shared pointer.

File: src/libs/costmap_2d/src/costmap_2d_ros.cpp (hunk 5 above)
Comment (bpwilcox) [reply src/libs/costmap_2d/src/costmap_2d_ros.cpp]: I made this an issue in #119

File: src/libs/costmap_2d/src/costmap_2d_ros.cpp (hunk 7 above)
Comment (bpwilcox) [reply src/libs/costmap_2d/src/costmap_2d_ros.cpp]: #119

**Pull request merged**